### PR TITLE
Improve docs build speed in canary runs

### PR DESCRIPTION
**Pull request by potiuk**
Self-hosted runners on ASF are SLOW to build docs (much slower than
regular public runners) but they have more disk space which is needed
to publish documentation. This PR splits docs building process to two
steps:

* First step that builds docs and uploads them as artifacts run on
  public runners

* Publish step that downloads the artifact and publishes them to
  S3 (this one needs more space and runs on ASF self-hosted runners)

This has the nice side effect that docs built in PR are available
as artifacts in PRs and main build.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by hussein-awala — APPROVED** [review]

File: .github/workflows/static-checks-mypy-docs.yml [hunk 1]
```diff
@@ -96,7 +96,7 @@ jobs:
   static-checks:
     timeout-minutes: 45
     name: "Static checks"
-    runs-on: ${{ fromJSON(inputs.runs-on-as-json-default) }}
+    runs-on: ${{ fromJSON(inputs.runs-on-) }}
```
Comment (shahar1) [inline .github/workflows/static-checks-mypy-docs.yml]: Seems like an accidental deletion:

```suggestion
    runs-on: ${{ fromJSON(inputs.runs-on-as-json-default) }}
```

**Comment by potiuk** [comment]
Right. :)

**Review by shahar1 — APPROVED** [review]

**Comment by shahar1** [comment]
Hmm, looks like it fails here:
https://github.com/apache/airflow/actions/runs/10221485186/job/28285050003?pr=41226

**Comment by potiuk** [comment]
> Hmm, looks like it fails here: https://github.com/apache/airflow/actions/runs/10221485186/job/28285050003?pr=41226

Yep. Just corrected it.

**Comment by potiuk** [comment]
It works now :)

**Pull request merged**